### PR TITLE
Profiles validations response

### PR DESCRIPTION
**Pull request by menabebawy**
- add profiles
- add messages.properties
- add more Junit tests to verify validation
- add DTO to Entities

### CODE REVIEW

File: api/src/main/java/com/mybooks/api/controller/AuthorController.java [hunk 1]
```diff
@@ -20,27 +22,28 @@ public AuthorController(AuthorService authorService) {
     }
 
     @GetMapping()
-    List<Author> getAllAuthors() {
-        return authorService.getAllAuthors();
+    ResponseEntity<List<AuthorDTO>> getAllAuthors() {
+        return new ResponseEntity<>(authorService.fetchAll(), HttpStatus.OK);
```
Comment (malirizk) [inline api/src/main/java/com/mybooks/api/controller/AuthorController.java]: Why this changes???
it should be

`@GetMapping()
@ResponseStatus(HttpStatus.OK)
    List<Author> getAllAuthors() {
        return authorService.getAllAuthors();
    }`

File: api/src/main/java/com/mybooks/api/configure/PropretiesConfig.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package com.mybooks.api.configure;
```
Comment (malirizk) [inline api/src/main/java/com/mybooks/api/configure/PropretiesConfig.java]: Could you refactor the package name from `configure` to `config`

File: api/src/main/java/com/mybooks/api/controller/AuthorControllerAdvice.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.web.bind.annotation.*;
+
+@RestControllerAdvice
+public class AuthorControllerAdvice {
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler(AuthorNotFoundException.class)
+    public ResponseEntity<String> resourceNotFoundException(AuthorNotFoundException ex) {
+        return new ResponseEntity<>("{ \"message\": \""+ex.getLocalizedMessage()+"\"}", HttpStatus.NOT_FOUND);
+    }
```
Comment (malirizk) [inline api/src/main/java/com/mybooks/api/controller/AuthorControllerAdvice.java]: `
String handleAuthorNotFoundException(AuthorNotFoundException ex) {
        return String.format("{ \"message\": \"{}\"}", ex.getLocalizedMessage());
    }
`

File: api/src/main/java/com/mybooks/api/controller/BookControllerAdvice.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@RestControllerAdvice
+public class BookControllerAdvice {
```
Comment (malirizk) [inline api/src/main/java/com/mybooks/api/controller/BookControllerAdvice.java]: You can create one class `ExceptionControllerAdvice`

File: api/src/main/java/com/mybooks/api/dto/BookDTO.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import javax.validation.constraints.NotNull;
+
+@Builder
+@Value
+public class BookDTO {
+    private String id;
+
+    @NotBlank(message = "{book.title.required}")
+    @NotNull(message = "{book.title.required}")
+    @NotEmpty(message = "{book.title.required}")
```
Comment (malirizk) [inline api/src/main/java/com/mybooks/api/dto/BookDTO.java]: I think @NotBlank is only sufficient, please read the docs

File: api/src/main/java/com/mybooks/api/service/AuthorService.java [hunk 6]
```diff
@@ -1,18 +1,18 @@
...
+    List<AuthorDTO> fetchAll();
 
-    Author getAuthorById(String id) throws AuthorNotFoundException;
+    AuthorDTO fetchById(String id) throws AuthorNotFoundException;
 
-    Author addAuthor(Author author);
+    AuthorDTO save(AuthorDTO authorDTO);
 
-    Author updateAuthor(Author updatedAuthor, String id) throws AuthorNotFoundException;
+    AuthorDTO update(AuthorDTO updatedAuthorDTO, String id) throws AuthorNotFoundException;
```
Comment (malirizk) [inline api/src/main/java/com/mybooks/api/service/AuthorService.java]: old methods names were better please revert

File: api/src/test/java/com/mybooks/api/controller/AuthorControllerTest.java [hunk 7]
```diff
@@ -24,30 +27,35 @@
 import java.util.List;
 import java.util.Objects;
 
+import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.mockito.Mockito.*;
 
 @WebMvcTest(AuthorController.class)
 public class AuthorControllerTest {
```
Comment (malirizk) [inline api/src/test/java/com/mybooks/api/controller/AuthorControllerTest.java]: remove public

File: api/src/test/java/com/mybooks/api/controller/AuthorControllerTest.java [hunk 8]
```diff
@@ -24,30 +27,35 @@
...
     ObjectMapper mapper;
-
     @MockBean
     AuthorServiceImpl authorService;
-    Author author1 = new Author("Author1_id", "Sam", "Simon");
-    Author author2 = new Author("Author2_id", "Adi", "John");
-    Author author3 = new Author("Author3_id", "Dodo", "Adel");
+    AuthorDTO author1 = AuthorDTO.builder().id("Author1_id").firstName("Sam").lastName("Simon").build();
+    AuthorDTO author2 = AuthorDTO.builder().id("Author2_id").firstName("Adi").lastName("John").build();
+    AuthorDTO author3 = AuthorDTO.builder().id("Author3_id").firstName("Dodo").lastName("Adel").build();
```
Comment (malirizk) [inline api/src/test/java/com/mybooks/api/controller/AuthorControllerTest.java]: could it be final?

File: api/src/test/java/com/mybooks/api/controller/AuthorControllerTest.java [hunk 9]
```diff
@@ -24,30 +27,35 @@
...
+    private AuthorMapper authorMapper;
+
+    @BeforeEach
+    void setUp() {
+        authorMapper = new AuthorMapperImpl();
+    }
 
     @Test
     public void getAllAuthors_success() throws Exception {
```
Comment (malirizk) [inline api/src/test/java/com/mybooks/api/controller/AuthorControllerTest.java]: public not needed

**Review by malirizk — COMMENTED** [review]
good job!!

**Comment by menabebawy** [comment]
@malirizk Thanks for your time putting into reviewing the PR, could you have a look?

**Review by malirizk — APPROVED** [review]
Looks good now
